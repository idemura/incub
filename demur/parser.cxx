#include "parser.hxx"
#include "grammar.tab.hxx"
#include <FlexLexer.h>

namespace igor {

bool parse(std::stringstream *ss) {
  yyFlexLexer lexer(ss);
  yy::parser parser(&lexer);
  int r = 0;
  while (!ss->eof() && r == 0) {
    r = parser.parse();
  }
  return r == 0;
}

}  // namespace
