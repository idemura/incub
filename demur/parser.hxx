#ifndef IGOR_PARSER_HXX
#define IGOR_PARSER_HXX

#include "base.hxx"

namespace igor {
bool parse(const string &file_name,
           std::function<void(const string&)> error_fn);
}  // namespace

#endif
