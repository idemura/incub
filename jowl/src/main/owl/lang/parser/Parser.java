/* Parser.java */
/* Generated By:JavaCC: Do not edit this line. Parser.java */
package owl.lang;

public class Parser implements ParserConstants {

  final public AstModule module() throws ParseException {AstModule module = new AstModule();
  AstFunction f;
    label_1:
    while (true) {
      switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
      case FN:{
        ;
        break;
        }
      default:
        jj_la1[0] = jj_gen;
        break label_1;
      }
      f = function();
module.addFunction(f);
    }
{if ("" != null) return module;}
    throw new Error("Missing return statement in function");
  }

  final public AstName qualifiedName() throws ParseException {AstName name = new AstName();
  Token tok;
    switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
    case NAME:{
      tok = jj_consume_token(NAME);
name.name = tok.image;
      label_2:
      while (true) {
        switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
        case 7:{
          ;
          break;
          }
        default:
          jj_la1[1] = jj_gen;
          break label_2;
        }
        jj_consume_token(7);
        tok = jj_consume_token(NAME);
name.name += "." + tok.image;
      }
{if ("" != null) return name;}
      break;
      }
    case 9:{
      jj_consume_token(9);
{if ("" != null) return AstName.WILDCARD;}
      break;
      }
    default:
      jj_la1[2] = jj_gen;
      jj_consume_token(-1);
      throw new ParseException();
    }
    throw new Error("Missing return statement in function");
  }

  final public AstFunction function() throws ParseException {AstFunction function = new AstFunction();
    jj_consume_token(FN);
    function.name = jj_consume_token(NAME).image;
    jj_consume_token(12);
    switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
    case NAME:
    case TYPE_NAME:{
      function.arguments = argumentList();
      break;
      }
    default:
      jj_la1[3] = jj_gen;
      ;
    }
    jj_consume_token(13);
    switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
    case 10:{
      jj_consume_token(10);
      function.type = type();
      break;
      }
    default:
      jj_la1[4] = jj_gen;
      ;
    }
    function.block = block();
{if ("" != null) return function;}
    throw new Error("Missing return statement in function");
  }

  final public AstArgumentList argumentList() throws ParseException {AstArgumentList argList = new AstArgumentList();
  AstArgument a;
    a = argument();
argList.add(a);
    label_3:
    while (true) {
      switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
      case 8:{
        ;
        break;
        }
      default:
        jj_la1[5] = jj_gen;
        break label_3;
      }
      jj_consume_token(8);
      a = argument();
argList.add(a);
    }
{if ("" != null) return argList;}
    throw new Error("Missing return statement in function");
  }

  final public AstArgument argument() throws ParseException {AstArgument argument = new AstArgument();
  Token tok;
    switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
    case NAME:{
      tok = jj_consume_token(NAME);
argument.name = tok.image;
      switch ((jj_ntk==-1)?jj_ntk_f():jj_ntk) {
      case 10:{
        jj_consume_token(10);
        argument.type = type();
        break;
        }
      default:
        jj_la1[6] = jj_gen;
        ;
      }
{if ("" != null) return argument;}
      break;
      }
    case TYPE_NAME:{
      argument.type = type();
{if ("" != null) return argument;}
      break;
      }
    default:
      jj_la1[7] = jj_gen;
      jj_consume_token(-1);
      throw new ParseException();
    }
    throw new Error("Missing return statement in function");
  }

  final public AstBlock block() throws ParseException {
    jj_consume_token(14);
    jj_consume_token(15);
{if ("" != null) return new AstBlock();}
    throw new Error("Missing return statement in function");
  }

  final public AstType type() throws ParseException {AstType type = new AstType();
  Token tok;
    tok = jj_consume_token(TYPE_NAME);
type.name = tok.image;
{if ("" != null) return type;}
    throw new Error("Missing return statement in function");
  }

  /** Generated Token Manager. */
  public ParserTokenManager token_source;
  SimpleCharStream jj_input_stream;
  /** Current token. */
  public Token token;
  /** Next token. */
  public Token jj_nt;
  private int jj_ntk;
  private int jj_gen;
  final private int[] jj_la1 = new int[8];
  static private int[] jj_la1_0;
  static {
      jj_la1_init_0();
   }
   private static void jj_la1_init_0() {
      jj_la1_0 = new int[] {0x40,0x80,0x8000200,0x18000000,0x400,0x100,0x400,0x18000000,};
   }

  /** Constructor with InputStream. */
  public Parser(java.io.InputStream stream) {
     this(stream, null);
  }
  /** Constructor with InputStream and supplied encoding */
  public Parser(java.io.InputStream stream, String encoding) {
    try { jj_input_stream = new SimpleCharStream(stream, encoding, 1, 1); } catch(java.io.UnsupportedEncodingException e) { throw new RuntimeException(e); }
    token_source = new ParserTokenManager(jj_input_stream);
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  /** Reinitialise. */
  public void ReInit(java.io.InputStream stream) {
     ReInit(stream, null);
  }
  /** Reinitialise. */
  public void ReInit(java.io.InputStream stream, String encoding) {
    try { jj_input_stream.ReInit(stream, encoding, 1, 1); } catch(java.io.UnsupportedEncodingException e) { throw new RuntimeException(e); }
    token_source.ReInit(jj_input_stream);
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  /** Constructor. */
  public Parser(java.io.Reader stream) {
    jj_input_stream = new SimpleCharStream(stream, 1, 1);
    token_source = new ParserTokenManager(jj_input_stream);
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  /** Reinitialise. */
  public void ReInit(java.io.Reader stream) {
    jj_input_stream.ReInit(stream, 1, 1);
    token_source.ReInit(jj_input_stream);
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  /** Constructor with generated Token Manager. */
  public Parser(ParserTokenManager tm) {
    token_source = tm;
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  /** Reinitialise. */
  public void ReInit(ParserTokenManager tm) {
    token_source = tm;
    token = new Token();
    jj_ntk = -1;
    jj_gen = 0;
    for (int i = 0; i < 8; i++) jj_la1[i] = -1;
  }

  private Token jj_consume_token(int kind) throws ParseException {
    Token oldToken;
    if ((oldToken = token).next != null) token = token.next;
    else token = token.next = token_source.getNextToken();
    jj_ntk = -1;
    if (token.kind == kind) {
      jj_gen++;
      return token;
    }
    token = oldToken;
    jj_kind = kind;
    throw generateParseException();
  }


/** Get the next Token. */
  final public Token getNextToken() {
    if (token.next != null) token = token.next;
    else token = token.next = token_source.getNextToken();
    jj_ntk = -1;
    jj_gen++;
    return token;
  }

/** Get the specific Token. */
  final public Token getToken(int index) {
    Token t = token;
    for (int i = 0; i < index; i++) {
      if (t.next != null) t = t.next;
      else t = t.next = token_source.getNextToken();
    }
    return t;
  }

  private int jj_ntk_f() {
    if ((jj_nt=token.next) == null)
      return (jj_ntk = (token.next=token_source.getNextToken()).kind);
    else
      return (jj_ntk = jj_nt.kind);
  }

  private java.util.List<int[]> jj_expentries = new java.util.ArrayList<int[]>();
  private int[] jj_expentry;
  private int jj_kind = -1;

  /** Generate ParseException. */
  public ParseException generateParseException() {
    jj_expentries.clear();
    boolean[] la1tokens = new boolean[29];
    if (jj_kind >= 0) {
      la1tokens[jj_kind] = true;
      jj_kind = -1;
    }
    for (int i = 0; i < 8; i++) {
      if (jj_la1[i] == jj_gen) {
        for (int j = 0; j < 32; j++) {
          if ((jj_la1_0[i] & (1<<j)) != 0) {
            la1tokens[j] = true;
          }
        }
      }
    }
    for (int i = 0; i < 29; i++) {
      if (la1tokens[i]) {
        jj_expentry = new int[1];
        jj_expentry[0] = i;
        jj_expentries.add(jj_expentry);
      }
    }
    int[][] exptokseq = new int[jj_expentries.size()][];
    for (int i = 0; i < jj_expentries.size(); i++) {
      exptokseq[i] = jj_expentries.get(i);
    }
    return new ParseException(token, exptokseq, tokenImage);
  }

  /** Enable tracing. */
  final public void enable_tracing() {
  }

  /** Disable tracing. */
  final public void disable_tracing() {
  }

  //
}
