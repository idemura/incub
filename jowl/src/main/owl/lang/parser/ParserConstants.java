/* Generated By:JavaCC: Do not edit this line. ParserConstants.java */
package owl.lang;


/**
 * Token literal values and constants.
 * Generated by org.javacc.parser.OtherFilesGen#start()
 */
interface ParserConstants {

  /**
   * End of File.
   */
  int EOF = 0;
  /**
   * RegularExpression Id.
   */
  int AS = 5;
  /**
   * RegularExpression Id.
   */
  int FN = 6;
  /**
   * RegularExpression Id.
   */
  int O = 18;
  /**
   * RegularExpression Id.
   */
  int D = 19;
  /**
   * RegularExpression Id.
   */
  int H = 20;
  /**
   * RegularExpression Id.
   */
  int NF = 21;
  /**
   * RegularExpression Id.
   */
  int TF = 22;
  /**
   * RegularExpression Id.
   */
  int L = 23;
  /**
   * RegularExpression Id.
   */
  int NUM_DEC = 24;
  /**
   * RegularExpression Id.
   */
  int NUM_HEX = 25;
  /**
   * RegularExpression Id.
   */
  int NUM_OCT = 26;
  /**
   * RegularExpression Id.
   */
  int NAME = 27;
  /**
   * RegularExpression Id.
   */
  int TYPE_NAME = 28;

  /**
   * Lexical state.
   */
  int DEFAULT = 0;

  /**
   * Literal token values.
   */
  String[] tokenImage = {
      "<EOF>",
      "\" \"",
      "\"\\r\"",
      "\"\\n\"",
      "\"\\t\"",
      "\"as\"",
      "\"fn\"",
      "\".\"",
      "\",\"",
      "\"_\"",
      "\":\"",
      "\";\"",
      "\"(\"",
      "\")\"",
      "\"{\"",
      "\"}\"",
      "\"[\"",
      "\"]\"",
      "<O>",
      "<D>",
      "<H>",
      "<NF>",
      "<TF>",
      "<L>",
      "<NUM_DEC>",
      "<NUM_HEX>",
      "<NUM_OCT>",
      "<NAME>",
      "<TYPE_NAME>",
  };

}
