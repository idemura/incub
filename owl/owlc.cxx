#include "parser.hxx"

FLAG_bool(help, false);

namespace igor {
namespace {

bool compile(const string &file) {
  ErrorLog elog(file);
  AST ast;
  if (parse(file, &ast, &elog)) {
    return ast.analyze(&elog);
  }
  return false;
}

}  // namespace
}  // namespace

int main(int argc, char **argv) {
  using namespace igor;
  std::ios_base::sync_with_stdio(false);
  if (!flags_parse(&argc, argv)) {
    return -1;
  }

  if (argc == 1 || flag_help) {
    cout<<"The Owl Language Compiler by Igor Demura\n"
          "igor.demura@gmail.com\n"
          "\n"
          "Usage:\n"
          "  owlc [file...]\n";
    return 0;
  }

  auto err_count = 0;
  for (int i = 1; i < argc; i++) {
    err_count += compile(argv[i]);
  }
  return err_count == 0? 0: -1;
}

