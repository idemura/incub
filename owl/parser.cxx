#include "parser.hxx"
#include "grammar.tab.hxx"

FLAG_bool(debug_bison, false);

namespace igor {

bool parse(const string &file_name, AST *ast, ErrorLog *elog) {
  auto res = false;
  auto f = fopen(file_name.c_str(), "rt");
  if (nullptr != f) {
    yyscan_t yyscanner;
    if (0 == yylex_init(&yyscanner)) {
      yyset_extra(elog, yyscanner);
      yyset_in(f, yyscanner);
#if YYDEBUG
      yydebug = flag_debug_bison;
#endif
      // Grammar accepts input till EOF.
      res = yyparse(yyscanner, ast, elog) == 0;
      yylex_destroy(yyscanner);
    } else {
      elog->error(0, 0).os()<<"internal error: scanner init";
      return false;
    }
    fclose(f);
  } else {
    elog->error(0, 0).os()<<"open file error: "<<file_name;
  }
  return res;
}

}  // namespace
