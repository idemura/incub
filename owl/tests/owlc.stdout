
INPUT: tests/fn_decl/0.owl
EXIT CODE: 255

INPUT: tests/fn_decl/10.owl
EXIT CODE: 255

INPUT: tests/fn_decl/11.owl
EXIT CODE: 255

INPUT: tests/fn_decl/12.owl
EXIT CODE: 255

INPUT: tests/fn_decl/13.owl
EXIT CODE: 255

INPUT: tests/fn_decl/14.owl
EXIT CODE: 255

INPUT: tests/fn_decl/15.owl
EXIT CODE: 255

INPUT: tests/fn_decl/1.owl
EXIT CODE: 255

INPUT: tests/fn_decl/2.owl
EXIT CODE: 255

INPUT: tests/fn_decl/3.owl
EXIT CODE: 255

INPUT: tests/fn_decl/4.owl
EXIT CODE: 255

INPUT: tests/fn_decl/5.owl
EXIT CODE: 255

INPUT: tests/fn_decl/6.owl
EXIT CODE: 255

INPUT: tests/fn_decl/7.owl
EXIT CODE: 255

INPUT: tests/fn_decl/8.owl
EXIT CODE: 255

INPUT: tests/fn_decl/9.owl
EXIT CODE: 255

INPUT: tests/fn_decl/basic.owl
EXIT CODE: 0

INPUT: tests/misc/basic.owl
EXIT CODE: 0

INPUT: tests/stmt/basic.owl
EXIT CODE: 0

INPUT: tests/type_decl/basic.owl
EXIT CODE: 0

